### ISSUE TITLE
Empty config even though `.env` is filled up

### ISSUE THREAD

**Issue by novito** [react-native-config/react-native-config#143]
I have created a react native app using [create-react-native-app](https://github.com/react-community/create-react-native-app) and I have this in my main `app.js`

```
import Config from 'react-native-config';
```

In my constructor I have:

```console.log(Config.API_KEY)```

And in my `.env` file I have:

```API_KEY = test```

That `Config.API_KEY` appears to be undefined.

Is this package supposed to work outside of Android and IOS (that is, straight on `app.js`)?

**Comment by rochapablo** [comment]
Same here.

**.env**
```
API_URL=http://
```

**.js**
```
import Config from 'react-native-config';
Config.API_URL // undefined
```

**Comment by rochapablo** [comment]
Problem fixed!

I haven't seen the [Setup](https://github.com/luggit/react-native-config#setup); Running ```react-native link react-native-config``` it worked.

**Comment by elyobo** [comment]
react-native link react-native-config didn't do everything for me, but
making the changes manually sorted it out

**Comment by hhsadiq** [comment]
having same issue. @elyobo how you fixed that?

**Comment by elyobo** [comment]
Carefully went through the instructions and checked that I'd done them all correctly.  I realised that the `react-native link react-native-config` hadn't made all of the expected changes, possibly because my MainApplication.java was already heavily modified to integrate `react-native-navigation`.  My recollection (possibly wrong) was that I had to manually go in and add `new ReactNativeConfigPackage()` to the last of packages returned in `MainApplication::getPackages()`.

**Comment by GramosBegolli** [comment]
How are you creating `.env` file?. This was happening to me when I was creating it from new Text Document, this does not recognize the `.env` extension. You should create the `.env` file by terminal. 

- Go to project root directory `cd Users/ProjectDir/`
- `touch .env` or e.x `touch .env.dev`
- Edit that file with text editor such as Atom

I think this should be described on the `Readme`

**Comment by hhsadiq** [comment]
Thanks guys, I was able to fix this issue. There were multiple issues with my project. I was using sudo in all of my commands and it caused some strange issues in auto generated files. Wasted lots of times yesterday and today eventually had to scrape my repo and re-clone it and build it from scratch without sudos. Now it's working. Also, there were issues in linking. First, I was using this plugin without running the link command. Yesterday figured it out and ran link command but since there were strange sudo based issues going on, so it did not work well.

So to summarize, use react-native commands without sudo and do run link commands. Hopefully, things will work.

**Comment by AlmogRnD** [comment]
I'm having the same issues, I followed all the steps for ios and it's still not working. I'm using WebStrom and running via terminal I have tried both `react-native run-ios --scheme CCMobileApp Staging` and `react-native run-ios `

**Comment by john1jan** [comment]
i am also facing the same issue. Config object is empty

**Comment by segalr3** [comment]
i have the same issue after upgrading to RN 0.47.
tried upgrading to react-native-config to 0.5.0 without luck.

**Comment by jurby** [comment]
+1 (react-native: 0.45.11, react: 16.0.0-alpha.12, react-native-config: ^0.6.0 / platform: android, development env: windows 10)

**Comment by john1jan** [comment]
I fixed it by unlinking and linking it again.

`react-native unlink react-native-config`
`react-native link react-native-config`

The only problem was i configured it wrongly before and linked it . And later even though fixing the configuration it was not working. Then unlinking and relinking worked

**Comment by AlmogRnD** [comment]
Tried unlink and link getting the following error 

ENOENT: no such file or directory, open '/Users/Almog/Full-Time-Employment/Construction-Cloud/RnD/CCMobileApp/android/app/src/main/java/com/ConstructionCloud/CCMobileApp/MainApplication.java'

**Comment by segalr3** [comment]
Tried to re-link. Also tried version 0.6.0.
I get no errors, but i always get back an empty object for Config.
rolling back to RN 0.44 and react-native-config 0.3.1 fixes it.

**Comment by DavidKongDesheng** [comment]
Delete the app from your device or simulator, reinstall and fixed~

**Comment by AlmogRnD** [comment]
@DavidKongDesheng that did not work, tried it twice. Everything works fine if you run it from Xcode but that kind of sucks prefer to run everything from terminal.

**Comment by amikofalvy** [comment]
I was able to get it working myself.
Make sure your env files are in the project root directory and not a subfolder.

Examine the ruby script:
```
  # find that above node_modules/react-native-config/ios/
  raw = File.read(File.join(Dir.pwd, "../../../#{file}"))
```

The location where `react-native-config` is looking for your .env* files is definitively in your project root directory.

Once I moved my config files to the correct location, I was able to access the config values.

Alternatively, for ios, if you want your files in a different directory you can modify the prebuild script to point to the directory you want to use.

I changed `echo ".env.dev" > /tmp/envfile` to `echo "app/.env.dev" > /tmp/envfile` so that I can put my config files in the `project-root/app` dir.

I hope this helps.

**Comment by Tim152** [comment]
the same issue

**Comment by swl367** [comment]
I'm having the same issue with v0.6.1. I've ran all my linking correctly (I see the xcode project with the library, etc). However, my Config from my import is an empty object.

**Comment by andrejunges** [comment]
The only solution for me was to remove `use_frameworks!` from podfile.
After removing that line and reinstalling my pods it started to work as expected

**Comment by MattHall** [comment]
Just dropping a note here to hopefully save someone from doing the same thing I did:

Don't put any spaces between the variable name, the equals and the value. Doing this:
```
VARIABLE_NAME = 'somevalue'
```
Will cause the `GeneratedInfoPlistDotEnv.h` file to silently exclude the value. Be sure that your .env looks like this:
```
VARIABLE_NAME='somevalue'
```

**Comment by karlingen** [comment]
+1 cannot get it to work. I get the same problem

I have tried running the link command, different simulators and doing everything stated in the Setup section in the Readme. But nothing works.

**Comment by karlingen** [comment]
I managed to solve this by running these commands inside my project: 

```
$ rm -rf $TMPDIR/react-* && watchman watch-del-all && rm -rf ios/build && rm -rf node_modules/ && npm cache clean --force && npm i
```

and then: 
```
$ react-native link react-native-config
```

**Comment by acharlop** [comment]
Adding the following line to `android/app/build.gradle` fixed it for me
`apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"`

**Comment by jayehmke** [comment]
I had the same issue, but @MattHall's solution fixed it for me.

**Comment by AstRonin** [comment]
I have a similar issue, when I create common release it works fine, and I see object in **Config** with params like **BUILD_TYPE**, **VERSION_CODE**, **VERSION_NAME** and my others from '.env' file ...
but
if I use **productFlavors** and run release with custom flavor then **Config** is empty object (**{}**) without any params.
Of cause, if I try get, for example, **Config.FB_KEY** I see undefined.
in flavors I change follow params: **applicationId**, **versionCode**, **versionName**, **signingConfig**, **manifestPlaceholders**.

What is problem can be?

**Comment by nlindroos** [comment]
An addition for (at least) React Native v0.50.4: `react-native link` did work for neither Android nor iOS. Android required the addition to `android/app/build.gradle` described by @acharlop. For iOS, `ReactNativeConfig.xcodeproj` needs to be manually added to _Libraries_ and `libReactNativeConfig.a` added to _Build Phases_ -> _Link Binary With Libraries_.

**Comment by francisrod01** [comment]
The same problem. I followed all steps in `Setup` too.

`.env.dev` file:

```
API_URL=http://192.168.0.107:3001/api/

```

`package.json` file:

```json
"scripts": {
  "android-dev": "ENVFILE=.env.dev react-native bundle --platform android --entry-file index.android.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res && react-native run-android",
    
```
* This config was fixed problem to load script from assets index.android.bundle.
See more about [here](https://stackoverflow.com/a/44476757/3332734)

`App.js` file:

```js
...

import Config from 'react-native-config'
....
render() {
    return (
        <View>
            <Text>Config is: {JSON.stringify(Config)}</Text>
        </View>
    )
}
....
```

**Screenshot:**

![image](https://user-images.githubusercontent.com/4691768/33904954-30bfc680-df64-11e7-800d-b23aa1bc521b.png)


What's going on?

**Comment by RubenO88** [comment]
My problem was that `react-native link` worked, but I already had multiple targets in Xcode and the libReactNativeConfig.a was not linked in my development environment. Thanks @nlindroos  Does anybody know how to add a new node package to a project and link it to both or possible even more targets?

**Comment by nlindroos** [comment]
@RubenO88 If `react-native link` doesn't do the necessary configuration automatically for a certain package, you can perform steps similar to the ones mentioned above, separately for each target. The required steps naturally differ between different packages.

**Comment by tmaly1980** [comment]
Guys, I think this only works when you relaunch a project. ie making modifications to the .env file doesn't take effect until you run the build all over again.

**Comment by rochapablo** [comment]
@tmaly1980, is right! You'll need re-build the project every time that your ```.env``` changes.

**Comment by zhangqiangoffice** [comment]
@MattHall ,thank you ! You saved my APP!

**Comment by 1oginov** [comment]
@elyobo thank you a lot! `react-native link` tried to use `MainApplication.java` that doesn't belongs to the project, so I've manually added `new ReactNativeConfigPackage()` like you mentioned. Just wanted to add that it shold be imported with `import com.lugg.ReactNativeConfig.ReactNativeConfigPackage;`

**Comment by bericp1** [comment]
For people running into this issue on Android using either `applicationIdSuffix` or build flavors, this is also a symptom of not properly setting the `build_config_package` `resValue` to your base package name (i.e. without the suffix) like the ["Advanced Android Setup"](https://github.com/luggit/react-native-config#advanced-android-setup) section in the readme describes. See also #15.

This is documented but may be hard to catch since the error message (`ReactNative: ReactConfig: Could not find BuildConfig class` logged in `android/src/main/java/com/lugg/ReactNativeConfig/ReactNativeConfigModule.java`) only shows up in logcat and doesn't explicitly cause a crash.

**Comment by BobAleena** [comment]
I'm seeing this issue when running npm test. Using react-native-config works when running my app though. 
I have a .env file in the root dir of the RN app which works. When running my test, it doesn't seem to find the .env file thus all the variables are undefined. I am assuming it is a matter of either creating another .env file for the tests specifically (but I don't know where to actually put that), or somehow specify the original .env file (also not sure how to do that).

**Comment by Tarpsvo** [comment]
Same issue for me. Works on iOS, but getting `BuildConfig` class not found error on Android.

```bash
D/ReactNative: ReactConfig: Could not find BuildConfig class
```

**Comment by andreleon** [comment]
On iOS I managed to fix this by changing the order of **Linked Frameworks dan Libraries**. Just drag it above `libReact.a`, recompile your project and it works.

![screen shot 2018-06-28 at 10 29 43](https://user-images.githubusercontent.com/1391492/42022724-4d68d95c-7abe-11e8-9da2-87617e630a37.png)

**Comment by ferlopez94** [comment]
This tutorial helped me: https://medium.com/differential/managing-configuration-in-react-native-cd2dfb5e6f7b.

**Comment by toeyanuntachai** [comment]
@elyobo Thank a lot,  I've manually added new ReactNativeConfigPackage(). 
it's worked

**Comment by stepankuzmin** [comment]
Same here.

```
"react": "16.6.0",
"react-native": "0.57.4",
"react-native-config": "0.11.5"
```

**Comment by luongs3** [comment]
each time i install this package, this issue shows up.
And i still don't know how to resolve it =))
Please, anyone ...

**Comment by RongGOHIN** [comment]
@luongs3 https://github.com/luggit/react-native-config#setup

**Comment by pacozaa** [comment]
Yes we should add `new ReactNativeConfigPackage()` but still doesn't solve the problem for me.

**Comment by pacozaa** [comment]
@bericp1 solotuion works for me. 
To conclude the solution 

1.  add `new ReactNativeConfigPackage()`
2.  read advance android setting and add as mentioned in document
```
defaultConfig {
    ...
    resValue "string", "build_config_package", "YOUR_PACKAGE_NAME_IN_ANDROIDMANIFEST.XML"
}
 
```

**Comment by tmaly1980** [comment]
I've dug into the module a bit more and how it works is the native code (Swift/Java) generates header files based on the contents of .env. Changes will not show up until the project is **cleaned and re-built**. The javascript code simply loads the constants defined in the native level code.

**Comment by jamessqr** [comment]
@nlindroos solution is still working and worked for me. I'm using React Native 0.57. I had to manually add the linked library in xcode and everything started working. This, despite the fact that react-native link command all reported back successful.

**Comment by slipkinem** [comment]
log: Could not find BuildConfig class. 
gradle version: 3.3.1
give up usage.

**Comment by sudiddo** [comment]
i manage to solve this issue by removing semicolon in the package json

this is the script for package.json in this repo's readme
> ENVFILE=".env.staging"; react-native run-ios

this is my package.json script

> "android": "ENVFILE=.env.dev react-native run-android"

i hope this help someone

**Comment by legshampoo** [comment]
I had this problem trying to use build variants for Android.  

Solved it by doing this in MainApplication.java
`import com.lugg.ReactNativeConfig.ReactNativeConfigPackage;`
and further down
`new ReactNativeConfigPackage()`

and then following the 'advanced android setup' step in the readme, but modifying it to be
`defaultConfig {
    resValue "string", "build_config_package", "${applicationId}"
}`
where the **${applicationId}** was the deciding factor in my case, because I was adding an 'applicationIdSuffix' to each build version

**Comment by The0racle** [comment]
If you're using multiple flavors on Android, with different `applicationId` (not `applicationIdSuffix`), do what @loginov-rocks and @bericp1 said as it should fix.

**Comment by conor909** [comment]
@legshampoo I always get this:
`ERROR: Project with path ':react-native-config' could not be found in project ':app'.`

Is there a linking step for Android?

**Comment by Chepkeitany** [comment]
Still getting the same issue. I get an empty object for `Config`. Works perfectly in the native code though only the Javascript part that's problematic.

**Comment by Chepkeitany** [comment]
@jeveloper That's a similar experience. Although I didn't take note when Config was being returned as an empty object. Now that I think about it, I think it was around that time when we switched to aab on the playstore.

**Comment by conor909** [comment]
I got it working, here's my setup:
`"react-native": "0.59.0"`
 `"react-native-config": "0.11.5"`

I'm also using yarn workspaces with the `.env` file outside the react-native project root, so to start android I need to give the path to the start command in my `package.json`:

```
scripts: {
   "start-android-dev": "ENVFILE=../../.env react-native run-android",
}
```

`> android/settings.gradle`
```
rootProject.name = 'MY-PROJECT-NAME'
...
include ':react-native-config'
project(':react-native-config').projectDir = new File(rootProject.projectDir, '/MY-PATH-TO/node_modules/react-native-config/android')
include ':app'
```

`> android/app/build.gradle`
```
apply plugin: "com.android.application"
apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"
...
...
project.ext.envConfigFiles = [
  debug : ".env",
  release : ".env",
  stagingrelease : ".env.staging",
  productionrelease : ".env.production"
]
...
...
dependencies {
    ...
    implementation project(':react-native-config')
    ...
}
```

`> android/app/main/src/java/MY-APP-NAME/MainApplication.java`
```
...
import com.lugg.ReactNativeConfig.ReactNativeConfigPackage;
import com.divesitedata.BuildConfig;
...
...
    @Override
    public boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
       ...
        new ReactNativeConfigPackage()
       ...
      );
    }
```

**Comment by helderburato** [comment]
Maybe in issue #400 I made a response that could help you.
See [here](https://github.com/luggit/react-native-config/issues/400#issuecomment-552002902).

**Comment by chenop** [comment]
Note that "PACKAGE_NAME_IN_ANDROID_MANIFEST.XML" means:
[Package name in **AndroidManifest.xml**]
not
[package name in AndroidManifest].xml

```
defaultConfig {
    resValue "string", "build_config_package", "com.packageName" // No .xml extension!
   ...
}
```

Yes I felt for it...

**Comment by TNChalise** [comment]
Mine was space at the beginning of the .env file.

**Comment by rubentlc** [comment]
In my case, what fixed the problem was this point: [ https://github.com/luggit/react-native-config#problems-with-proguard]( https://github.com/luggit/react-native-config#problems-with-proguard)

**Comment by AizenSousuke** [comment]
is there a fix for this issue at this point of time? doesn't seem to work in js

**Comment by dylanhillier** [comment]
I had some issues recently on iOS where changes in my .env file were not reflected in my mobile app.
What didn't work:

- deleting the ios/build directory and rebuilding.
- running pod install again, then rebuilding.
- clearing metro cache and rebuilding.
- deleting the app from the simulator and rebuilding + reinstall.

What did work:

- Close the simulator + close metro + rebuild app.

It's not entirely clear whether or not the need to close the simulator is a result of some misconfiguration on my end.
I'm using react-native-config@1.5.0.

**Comment by owen800q** [comment]
@dylanhillier 
Working for me, I think the doc need to mention every time anything inside .env has been changed, need to rebuild the app.

**Comment by dylanhillier** [comment]
@owen800q 

I use the scripts from within `package.json` typically to build/run my react-native mobile app.

On Android, if my mobile app is running via a simulator, and I make a change to the `.env` file, then for this change to be make available to my app, i just need to re-run `yarn android`. I consider this to be working as expected.

On iOS, if my mobile app is running via a simulator, and i make a change to the `.env` file, then this change **_will not be visible_** to my mobile app by simply running `yarn ios`. I consider this to be not working as expected. For my `.env` changes to be available, I have to run the build via xcode.

This is inconsistent and undesirable. It would be nice for it to just work via `yarn ios` and to be consistent between ios and android.

**Comment by Biplovkumar** [comment]
Same issue in `1.5.1.`

**Comment by luancurti** [comment]
Closing this as superseded, and I want to be straight about what that means rather than claim a single fix.

This thread was opened in 2017 against `create-react-native-app` and a version of the library that no longer resembles the current one. Over nine years it became the place people landed whenever `Config` came out empty, which meant it collected several unrelated causes in one place — and a thread with several causes can never actually be resolved. Each of those causes has since been addressed on its own:

- **`Config` logs as `{"getConstants": [Function], "toString": [Function]}`, or values are `undefined` on the New Architecture.** The entry point was `NativeModules.RNCConfigModule || {}`, a lazy proxy whose constants are never spread onto the object. Rewritten to call `getConfig()` explicitly in **v1.6.0**.
- **Empty on Android with `applicationIdSuffix` or a per-flavor `applicationId`.** `BuildConfig` is generated in the module's `namespace`, but was only ever looked up via the applicationId, so those variants missed it and JS received `{}`. Fixed in **v1.6.2** ([#879](https://github.com/react-native-config/react-native-config/pull/879)); resolution now walks candidates and logs every package it tried when all of them miss.
- **The native module is not registered at all** — not rebuilt after install, autolinking disabled, or manually linked on the New Architecture. This used to be swallowed by that `|| {}` and surface as an empty object. **v1.6.2** ([#880](https://github.com/react-native-config/react-native-config/pull/880)) replaced it with an error naming each likely cause.
- **iOS: no `.env` was found, or the codegen build phase never ran.** The build wrote an empty `DOT_ENV` and succeeded, leaving `{}` with no diagnostic anywhere. **v1.7.0** now names the path it looked for, at build time and at runtime.
- **Proguard stripping `BuildConfig` in Android release builds** — covered in [Troubleshooting](https://github.com/react-native-config/react-native-config#problems-with-proguard).

The common thread was that an empty `Config` explained nothing about itself. On **[v1.7.0](https://github.com/react-native-config/react-native-config/releases/tag/v1.7.0)** every path above either fails loudly or says which file it read and where it looked, so a fresh report can be diagnosed instead of joining a queue behind nine years of unrelated ones.

That is the reason for closing: not that every comment here was individually fixed, but that this thread can no longer distinguish which of them anyone is hitting. It is being locked with the rest of the batch.

**If you are still affected on 1.7.0**, please [open a new issue](https://github.com/react-native-config/react-native-config/issues/new/choose) — it will get a real answer rather than a subscription to this one. Include your platform, whether the New Architecture is enabled, and any `[react-native-config]` line from the console or `ENVFILE=` line from the build log. Those now say what went wrong, and with them a fix is usually quick.

Thanks to everyone who debugged in here over the years — a lot of what shipped in 1.6.x and 1.7.0 came from this thread.
